nvc.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "NVC" "1" "December 2013" "" "NVC Manual"
.
.SH "NAME"
\fBnvc\fR \- VHDL Compiler and Simulator
.
.SH "SYNOPSIS"
\fBnvc\fR \-a [\fIoptions\fR] \fIfiles\fR\.\.\.
.
.br
\fBnvc\fR \-e [\fIoptions\fR] \fIunit\fR
.
.br
\fBnvc\fR \-r [\fIoptions\fR] \fIunit\fR
.
.br
.
.SH "DESCRIPTION"
NVC is an implementation of the VHDL language as defined by IEEE standard 1076\-1993 and later revisions\. Simulating a design typically involes three steps: \fBanalysing\fR one more more source files into the work library; \fBelaborating\fR a top\-level design unit; and \fBrunning\fR the elaborated design\.
.
.SH "OPTIONS"
NVC accepts three kinds of options: global options; a single command option; and options specific to the command\. Global options must be placed before the command and specific options must be placed after the command\.
.
.SS "Commands"
.
.TP
\fB\-a\fR \fIfiles\fR
Analyse \fIfiles\fR into the work library\.
.
.TP
\fB\-e\fR \fIunit\fR
Elaborate a previously analysed top level design unit\.
.
.TP
\fB\-r\fR \fIunit\fR
Execute a previously elaborated top level design unit\.
.
.TP
\fB\-\-codegen\fR \fIunit\fR
Generate a native shared library for a previously analysed package\. This can improve runtime performance if the package contains a large number of frequently used subprograms\.
.
.TP
\fB\-\-dump\fR \fIunit\fR
Print out a pseudo\-VHDL representation of an analysed unit\. This is usually only useful for debugging the compiler\.
.
.SS "Global options"
.
.TP
\fB\-h\fR, \fB\-\-help\fR
Display usage summary\.
.
.TP
\fB\-L\fR \fIpath\fR
Add \fIpath\fR to the list of directories to search for libraries (see \fILIBRARIES\fR section below)\.
.
.TP
\fB\-v\fR, \fB\-\-version\fR
Display version and copyright information\.
.
.TP
\fB\-\-work=\fR \fIname\fR
Use \fIname\fR as the work library (see \fILIBRARIES\fR section below)\.
.
.SS "Analysis options"
.
.TP
\fB\-\-bootstrap\fR
Allow compilation of the STANDARD package\. Not intended for end users\.
.
.TP
\fB\-\-prefer\-explicit\fR
Any visible explicitly declared operator always hides an implicit operator regardless of the region in which it is declared\. This deviates from the VHDL standard but is required to analyse code that uses the Synopsys \fBstd_logic_arith\fR package\.
.
.SS "Elaboration options"
.
.TP
\fB\-\-cover\fR
Enable code coverage reporting (see the \fICODE COVERAGE\fR section below)\.
.
.TP
\fB\-\-disable\-opt\fR
Disable LLVM optimisations\. Not generally useful unless debugging the generated LLVM IR\.
.
.TP
\fB\-\-dump\-llvm\fR
Print generated LLVM IR prior to optimisation\.
.
.TP
\fB\-\-native\fR
Generate native code shared library\. By default NVC will use LLVM JIT compilation to generate machine code at runtime\. For large designs compiling to native code at elaboration time may improve performance\.
.
.SS "Runtime options"
.
.TP
\fB\-b\fR, \fB\-\-batch\fR
Run in batch mode\. This is the default\.
.
.TP
\fB\-c\fR, \fB\-\-command\fR
Run in interactive TCL command line mode\. See \fITCL SHELL\fR section below\.
.
.TP
\fB\-\-format=\fR\fIfmt\fR
Generate waveform data in format \fIfmt\fR\. Currently supported formats are: \fBfst\fR, \fBlxt\fR, and \fBvcd\fR\. The FST and LXT formats are native to GtkWave\. The FST format is preferred over LXT due its smaller size and better performance; however VHDL support in FST requires a recent version of GtkWave so LXT is provided for compatibility\. VCD is a very widely used format but has limited ability to represent VHDL types and the performance is poor: select this only if you must use the output with a tool that does not support FST or LXT\. The default format is FST if this option is not provided\.
.
.TP
\fB\-\-stats\fR
Print time and memory statistics at the end of the run\.
.
.TP
\fB\-\-stop\-delta=\fR\fIN\fR
Stop after \fIN\fR delta cycles\. This can be used to detect zero\-time loops in your model\. The default is 1000 if not specified\. Setting this to zero disables the delta cycle limit\.
.
.TP
\fB\-\-stop\-time=\fR\fIT\fR
Stop the simulation after the given time has elapsed\. Format of \fIT\fR is an integer followed by a time unit in lower case\. For example \fB5ns\fR or \fB20ms\fR\.
.
.TP
\fB\-\-trace\fR
Trace simulation events\. This is usually only useful for debugging the simulator\.
.
.TP
\fB\-w, \-\-wave=\fR\fIfile\fR
Write waveform data to \fIfile\fR\. The file name is optional and if not specified will default to the name of the top\-level unit with the appropriate extension for the waveform format\. The waveform format can be specified with the \fB\-\-format\fR option\.
.
.SH "ENVIRONMENT VARIABLES"
.
.TP
\fBNVC_CYG_LIB\fR:

.
.TP
\fBNVC_DECODE\fR:

.
.TP
\fBNVC_EVAL_DEBUG\fR:

.
.TP
\fBNVC_GC_VERBOSE\fR:

.
.TP
\fBNVC_GDB\fR:

.
.TP
\fBNVC_LIBPATH\fR:

.
.TP
\fBNVC_LINK_QUIET\fR:

.
.TP
\fBNVC_LLC_ARG\fR:

.
.TP
\fBNVC_NO_COLOR\fR:

.
.TP
\fBNVC_TREE_SIZES\fR:

.
.SH "LIBRARIES"
Description of library search path, contents, etc\.
.
.SH "CODE COVERAGE"
Description of coverage generation
.
.SH "TCL SHELL"
Describe interactive TCL shell
.
.SH "AUTHOR"
Written by Nick Gasson
.
.SH "REPORTING BUGS"
Report bugs using the GitHub issue tracker at
.
.br
\fIhttps://github\.com/nickg/nvc/issues\fR


src/nvc.c
//
//  Copyright (C) 2011-2012  Nick Gasson
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

#include "util.h"
#include "parse.h"
#include "phase.h"
#include "rt/rt.h"
#include "rt/slave.h"

#include <unistd.h>
#include <getopt.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <assert.h>
#if defined HAVE_TCL_TCL_H
#include <tcl/tcl.h>
#elif defined HAVE_TCL_H
#include <tcl.h>
#endif

static const char *work_name = "work";

static void set_work_lib(void)
{
   lib_t work = lib_find(work_name, false, false);
   if (work == NULL) {
      if ((work = lib_new(work_name)) == NULL)
         exit(EXIT_FAILURE);
   }

   lib_set_work(work);
}

static ident_t to_unit_name(const char *str)
{
   char *name;
   char *p;
   ident_t i;

   name = strdup(str);
   for (p = name; *p; p++)
      *p = toupper((uint8_t)*p);

   i = ident_prefix(lib_name(lib_work()),
                    ident_new(name), '.');
   free(name);
   return i;
}

static int analyse(int argc, char **argv)
{
   static struct option long_options[] = {
      {"bootstrap", no_argument, 0, 'b'},
      {"dump-llvm", no_argument, 0, 'd'},
      {0, 0, 0, 0}
   };
   int c, index = 0;
   const char *spec = "";
   tree_t to_cgen[16];
   int n_cgen = 0;
   int i;
   tree_t unit;

   set_work_lib();

   optind = 1;
   while ((c = getopt_long(argc, argv, spec, long_options, &index)) != -1) {
      switch (c) {
      case 0:
         // Set a flag
         break;
      case '?':
         // getopt_long already printed an error message
         exit(EXIT_FAILURE);
      case 'b':
         opt_set_int("bootstrap", 1);
         break;
      case 'd':
         opt_set_int("dump-llvm", 1);
         break;
      default:
         abort();
      }
   }

   for (i = optind; i < argc; i++) {
      if (!input_from_file(argv[i]))
         return EXIT_FAILURE;

      while ((unit = parse())) {
         if (sem_check(unit)) {
            assert(sem_errors() == 0);
            unalias(unit);
            simplify(unit);

            if (tree_kind(unit) == T_PACK_BODY) {
               assert(n_cgen < ARRAY_LEN(to_cgen));
               to_cgen[n_cgen++] = unit;
            }
         }
         else
            break;
      }

      tree_gc();

      if (parse_errors() + sem_errors() + simplify_errors() > 0)
         return EXIT_FAILURE;

      lib_save(lib_work());
   }

   for (i = 0; i < n_cgen; i++)
      cgen(to_cgen[i]);

   return EXIT_SUCCESS;
}

static int elaborate(int argc, char **argv)
{
   static struct option long_options[] = {
      {"disable-opt", no_argument, 0, 'o'},
      {"dump-llvm", no_argument, 0, 'd'},
      {"native", no_argument, 0, 'n'},
      {0, 0, 0, 0}
   };
   int c, index = 0;
   const char *spec = "";
   ident_t unit_i;
   tree_t unit, e;

   set_work_lib();

   optind = 1;
   while ((c = getopt_long(argc, argv, spec, long_options, &index)) != -1) {
      switch (c) {
      case 'o':
         opt_set_int("optimise", 0);
         break;
      case 'd':
         opt_set_int("dump-llvm", 1);
         break;
      case 'n':
         opt_set_int("native", 1);
         break;
      case 0:
         // Set a flag
         break;
      case '?':
         // getopt_long already printed an error message
         exit(EXIT_FAILURE);
      default:
         abort();
      }
   }

   if (optind == argc)
      fatal("missing top-level unit name");

   unit_i = to_unit_name(argv[optind]);
   unit = lib_get(lib_work(), unit_i);
   if (unit == NULL)
      fatal("cannot find unit %s in library %s",
            istr(unit_i), istr(lib_name(lib_work())));

   e = elab(unit);
   if (e == NULL)
      return EXIT_FAILURE;

   opt(e);

   tree_gc();

   // Save the library now so the code generator can attach temporary
   // meta data to trees. We also want to avoid saving the signal
   // driver lists which will be very large
   lib_save(lib_work());

   driver_extract(e);
   if (driver_errors() > 0)
      return EXIT_FAILURE;

   cgen(e);
   link_bc(e);

   return EXIT_SUCCESS;
}

static uint64_t parse_time(const char *str)
{
   char     unit[4];
   unsigned base;
   uint64_t mult = 1;

   if (sscanf(str, "%u%3s", &base, unit) != 2)
      fatal("invalid time format: %s", str);

   if      (strcmp(unit, "fs") == 0) mult = 1;
   else if (strcmp(unit, "ps") == 0) mult = 1000;
   else if (strcmp(unit, "ns") == 0) mult = 1000000;
   else if (strcmp(unit, "us") == 0) mult = 1000000000;
   else if (strcmp(unit, "ms") == 0) mult = 1000000000000;
   else
      fatal("invalid unit: %s", unit);

   return base * mult;
}

static int run(int argc, char **argv)
{
   static struct option long_options[] = {
      {"trace", no_argument, 0, 't'},
#if defined HAVE_TCL_TCL_H || defined HAVE_TCL_H
      {"batch", no_argument, 0, 'b'},
      {"command", no_argument, 0, 'c'},
#endif
      {"stop-time", required_argument, 0, 's'},
      {"vcd", required_argument, 0, 'v'},
      {"stats", no_argument, 0, 'S'},
      {0, 0, 0, 0}
   };
#if defined HAVE_TCL_TCL_H || defined HAVE_TCL_H
   enum { BATCH, COMMAND } mode = BATCH;
#endif
   uint64_t stop_time = UINT64_MAX;
   const char *vcd_fname = NULL;
   int c, index = 0;
   const char *spec = "bc";
   ident_t top, ename;
   tree_t e;
   tree_rd_ctx_t ctx;

   set_work_lib();

   optind = 1;
   while ((c = getopt_long(argc, argv, spec, long_options, &index)) != -1) {
      switch (c) {
      case 0:
         // Set a flag
         break;
      case '?':
         // getopt_long already printed an error message
         exit(EXIT_FAILURE);
      case 't':
         opt_set_int("rt_trace_en", 1);
         break;
#if defined HAVE_TCL_TCL_H || defined HAVE_TCL_H
      case 'b':
         mode = BATCH;
         break;
      case 'c':
         mode = COMMAND;
         break;
#endif
      case 's':
         stop_time = parse_time(optarg);
         break;
      case 'v':
         vcd_fname = optarg;
         break;
      case 'S':
         opt_set_int("rt-stats", 1);
         break;
      default:
         abort();
      }
   }

   if (optind == argc)
      fatal("missing top-level unit name");

   top = to_unit_name(argv[optind]);
   ename = ident_prefix(top, ident_new("elab"), '.');
   e = lib_get_ctx(lib_work(), ename, &ctx);
   if (e == NULL)
      fatal("%s not elaborated", istr(top));
   else if (tree_kind(e) != T_ELAB)
      fatal("%s not suitable top level", istr(top));

   if (vcd_fname != NULL)
      vcd_init(vcd_fname, e);

#if defined HAVE_TCL_TCL_H || defined HAVE_TCL_H
   if (mode == BATCH)
      rt_batch_exec(e, stop_time, ctx);
   else {
      bool master = slave_fork();
      if (master)
         shell_run(e);
      else
         rt_slave_exec(e, ctx);
   }
#else
   rt_batch_exec(e, stop_time, ctx);
#endif

   tree_read_end(ctx);
   return EXIT_SUCCESS;
}

static int dump_cmd(int argc, char **argv)
{
   static struct option long_options[] = {
      {"elab", no_argument, 0, 'e'},
      {"body", no_argument, 0, 'b'},
      {0, 0, 0, 0}
   };
   bool add_elab = false, add_body = false;
   int c, index = 0;
   const char *spec = "eb";
   int i;
   ident_t name;
   tree_t top;

   set_work_lib();

   optind = 1;
   while ((c = getopt_long(argc, argv, spec, long_options, &index)) != -1) {
      switch (c) {
      case 0:
         // Set a flag
         break;
      case '?':
         // getopt_long already printed an error message
         exit(EXIT_FAILURE);
      case 'e':
         add_elab = true;
         break;
      case 'b':
         add_body = true;
         break;
      default:
         abort();
      }
   }

   if (optind == argc)
      fatal("missing unit name");

   for (i = optind; i < argc; i++) {
      name = to_unit_name(argv[i]);
      if (add_elab)
         name = ident_prefix(name, ident_new("elab"), '.');
      else if (add_body)
         name = ident_prefix(name, ident_new("body"), '-');
      top = lib_get(lib_work(), name);
      if (top == NULL)
         fatal("%s not analysed", istr(name));
      dump(top);
   }

   return EXIT_SUCCESS;
}

static void set_default_opts(void)
{
   opt_set_int("rt-stats", 0);
   opt_set_int("rt_trace_en", 0);
   opt_set_int("dump-llvm", 0);
   opt_set_int("optimise", 1);
   opt_set_int("native", 0);
   opt_set_int("bootstrap", 0);
}

static void usage(void)
{
   printf("Usage: %s [OPTION]... COMMAND [OPTION]...\n"
          "\n"
          "COMMAND is one of:\n"
          " -a [OPTION]... FILE...\tAnalyse FILEs into work library\n"
          " -e UNIT\t\tElaborate and generate code for UNIT\n"
          " -r UNIT\t\tExecute previously elaborated UNIT\n"
          " --dump UNIT\t\tPrint out previously analysed UNIT\n"
          "\n"
          "Global options may be placed before COMMAND:\n"
          " -v, --version\t\tDisplay version and copyright information\n"
          " -h, --help\t\tDisplay this message and exit\n"
          "     --work=NAME\tUse NAME as the work library\n"
          "\n"
          "Analyse options:\n"
          "     --bootstrap\tAllow compilation of STANDARD package\n"
          "\n"
          "Elaborate options:\n"
          "     --disable-opt\tDisable LLVM optimisations\n"
          "     --dump-llvm\tPrint generated LLVM IR\n"
          "     --native\t\tGenerate native code shared library\n"
          "\n"
          "Run options:\n"
#if defined HAVE_TCL_TCL_H || defined HAVE_TCL_H
          " -b, --batch\t\tRun in batch mode (default)\n"
          " -c, --command\t\tRun in TCL command line mode\n"
#endif
          "     --stats\t\tPrint statistics at end of run\n"
          "     --stop-time=T\tStop after simulation time T (e.g. 5ns)\n"
          "     --trace\t\tTrace simulation events\n"
          "     --vcd=FILE\t\tWrite VCD data to FILE\n"
          "\n"
          "Dump options:\n"
          " -e, --elab\t\tDump an elaborated unit\n"
          " -b, --body\t\tDump package body\n"
          "\n"
          "Report bugs to %s\n",
          PACKAGE, PACKAGE_BUGREPORT);
}

static void version(void)
{
   static const char *copy =
      "Copyright (C) 2011-2012  Nick Gasson\n"
      "This program comes with ABSOLUTELY NO WARRANTY. This is free software, and\n"
      "you are welcome to redistribute it under certain conditions. See the GNU\n"
      "General Public Licence for details.";

#ifdef HAVE_CONFIG_H
#if defined HAVE_TCL_TCL_H || defined HAVE_TCL_H
   printf("%s (llvm %s; tcl %s)\n",
          PACKAGE_STRING, LLVM_VERSION, TCL_VERSION);
#else
   printf("%s (llvm %s)\n",
          PACKAGE_STRING, LLVM_VERSION);
#endif
#endif

   puts(copy);
}

int main(int argc, char **argv)
{
   static struct option long_options[] = {
      {"help",    no_argument,       0, 'h'},
      {"version", no_argument,       0, 'v'},
      {"work",    required_argument, 0, 'w'},
      {"dump",    no_argument,       0, 'd'},
      {0, 0, 0, 0}
   };
   int c, index = 0;
   const char *spec = "aehr";

   term_init();
   register_trace_signal_handlers();
   set_default_opts();

   while ((c = getopt_long(argc, argv, spec, long_options, &index)) != -1) {
      switch (c) {
      case 0:
         // Set a flag
         break;
      case 'h':
         usage();
         exit(EXIT_SUCCESS);
      case 'v':
         version();
         exit(EXIT_SUCCESS);
      case 'w':
         work_name = optarg;
         break;
      case 'a':
      case 'e':
      case 'd':
      case 'r':
         // Subcommand options are parsed later
         argc -= (optind - 1);
         argv += (optind - 1);
         goto getopt_out;
      case '?':
         // getopt_long already printed an error message
         exit(EXIT_FAILURE);
      default:
         abort();
      }
   }
 getopt_out:

   switch (c) {
   case 'a':
      return analyse(argc, argv);
   case 'e':
      return elaborate(argc, argv);
   case 'r':
      return run(argc, argv);
   case 'd':
      return dump_cmd(argc, argv);
   default:
      fprintf(stderr, "%s: missing command\n", PACKAGE);
      return EXIT_FAILURE;
   }
}
